Redesign of the scripts

omron93 opened the pull request:
For 2.6 only. I will do the same for 2.4 after 2.6 will be ready for merge!

Additional refactoring of the scripts.
- run-mongod divided into several parts - pre-init (checking variables,...), running mongod on localhost, init (creating users,...), post-init (before execing mongod), exec mongod

Reworked replication too.
- no need to run one pod only for initiation and stopping it
- one pod run with command "run-mongod-replication initiation", other pods with command "run-mongod-replication" will connect it...

I've tested it only on localhost, so it probably won't work in this state - I am really not sure how the config .json should look like. @mfojtik @bparees Could someone please help me with this?

Commit:
Redesign of the scripts

omron93 commented:
No README changes yet too... will be added later.

rhcarvalho commented:
At a quick glance this looks like a major rewrite without accompanying automated tests, which poses too big a risk to merge as is now.

@omron93 how about breaking this into smaller chunks that are 1) easier to verify correctness 2) easier to test?

Before adding `2.6/examples/replica/mongodb-clustered.json` I would like us to revisit the implementation of the example for 2.4 -- or, at least, implement a better replication example that we could eventually back-port to 2.4. It seems to me that, except for being in different files, the replication code and architecture is essentially the same as we see today in 2.4?

rhcarvalho commented:
We can merge the changes not related to replication.
The part related to replication depends on adding extended tests to OpenShift that would test  replication scenarios; those tests don't exist yet :cry: 

@omron93 do you mind splitting this PR into two, and the replication part for 2.6 will go separate?
We want replication in 2.6, but we need it tested to avoid regressions we've faced with 2.4.

omron93 commented:
> At a quick glance this looks like a major rewrite without accompanying automated tests, which poses too big a risk to merge as is now.

Single mongod (no replication) passes the same tests normally! So tests are not only for replication part and this is the same situation as for 2.4 now...

> Before adding 2.6/examples/replica/mongodb-clustered.json I would like us to revisit the implementation of the example for 2.4 -- or, at least, implement a better replication example that we could eventually back-port to 2.4. It seems to me that, except for being in different files, the replication code and architecture is essentially the same as we see today in 2.4?

No, pieces of code are the same, but architecture is really different. There is completely different process of initializing replication and adding/removing members of replicaset. It is because I am really not sure that current implementation which all pods have to be ready to initialize replset is the best solution... so changing code in 2.4 is the same risk as in 2.6 ... there are no tests for both cases!

> do you mind splitting this PR into two, and the replication part for 2.6 will go separate?
> We want replication in 2.6, but we need it tested to avoid regressions we've faced with 2.4.

The code for no replication is quite the same (I hope more clean and easier to maintain) and main rework is in replication part. It is not easy to split it - in current implementation non-replication and replication code is mixed in run-mongod script... I think it is waste of time to only split it and change is not so big.
Also you've said that replication is not supported in 2.6 yet, so why not to start to rework it to have it better and when there will be some replication tests, you can test it and start to support replication in 2.6.

@rhcarvalho What is the current state of tests for replication?

mfojtik commented:
@omron93 there are no extended tests for mongodb replication, which is why we hesitate to add replication support to mongo 2.6. We will start working on them soon (after we finish the trello cards for this sprint). Can this PR be hold until we have the tests? We really don't want to break users that will be deploying the replication template with mongodb 2.6.

omron93 commented:
@mfojtik So lets wait with this PR...

rhcarvalho commented:
FYI I'm still tracking this. Getting the replication example using Mongo 2.6 is a priority.

@omron93, I'll manually test this in OpenShift when I get time and let you know the outcome.
Have you tried that as well?

Meanwhile we're also working on automated tests that will help us greatly.

omron93 commented:
> I'll manually test this in OpenShift when I get time and let you know the outcome.
> Have you tried that as well?

Thanks. I've test it only on local host (with docker command, with little bit modified endpoints() function). I am really not sure about the .json config for replication. @rhcarvalho Please could you review this first and maybe correct it to allow me  to test it in OpenShift too.

rhcarvalho commented:
@omron93 I've fetched this PR to test the changes, but was unable to build the images locally because https://www.softwarecollections.org/ is down.

```
$ make build VERSION=2.6
...
Cannot open: https://www.softwarecollections.org/en/scls/rhscl/v8314/epel-7-x86_64/download/rhscl-v8314-epel-7-x86_64.noarch.rpm. Skipping.
Cannot open: https://www.softwarecollections.org/en/scls/rhscl/rh-mongodb26/epel-7-x86_64/download/rhscl-rh-mongodb26-epel-7-x86_64.noarch.rpm. Skipping.
Cannot open: https://www.softwarecollections.org/en/scls/rhscl/rh-java-common/epel-7-x86_64/download/rhscl-rh-java-common-epel-7-x86_64.noarch.rpm. Skipping.
Error: Nothing to do
The command '/bin/sh -c yum install -y epel-release &&     rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 &&     yum -y --setopt=tsflags=nodocs install     https://www.softwarecollections.org/en/scls/rhscl/v8314/epel-7-x86_64/download/rhscl-v8314-epel-7-x86_64.noarch.rpm     https://www.softwarecollections.org/en/scls/rhscl/rh-mongodb26/epel-7-x86_64/download/rhscl-rh-mongodb26-epel-7-x86_64.noarch.rpm     https://www.softwarecollections.org/en/scls/rhscl/rh-java-common/epel-7-x86_64/download/rhscl-rh-java-common-epel-7-x86_64.noarch.rpm &&     yum install -y --setopt=tsflags=nodocs bind-utils gettext iproute v8314 rh-mongodb26-mongodb rh-mongodb26 &&     yum clean all &&     mkdir -p /var/lib/mongodb/data && chown -R mongodb.0 /var/lib/mongodb/ /var/opt/rh/rh-mongodb26/lib/mongodb &&     chmod g+w -R /var/opt/rh/rh-mongodb26/lib/mongodb &&     chmod -R g+rwx /var/lib/mongodb' returned a non-zero code: 1
hack/common.mk:21: recipe for target 'build' failed
make: *** [build] Error 1
```

rhcarvalho commented:
Since https://www.softwarecollections.org/ is still down, I'll need to postpone and test this ~2 weeks later.
@omron93 by that time I'd like to pair with you and we can do it together :sunglasses:

omron93 commented:
Thanks.

Late for you, but https://www.softwarecollections.org/ is back online

omron93 commented:
Any progress in OpenShift replication tests? And things connected with it?

We are preparing container image fro rh-mongodb32 so it would be very good to be able to do it better... ;-)

rhcarvalho commented:
> Any progress in OpenShift replication tests?

@omron93 we've only managed to get automated tests for the standalone scenario merged in openshift/origin, no replication tests yet, but @php-coder is working on it.

I'm still willing to meet you and do the manual testing together ;)

rhcarvalho commented:
@omron93 reminders after our conversation:
- [ ] need to make `2.6/root/usr/bin/run-mongod-replication` terminate at some point (because the deployment only finishes after the post hook command finishes)
- [ ] update initial number of replicas to 3 to match the example in `2.4`
- [ ] don't need "container-entrypoint" in the post hook command

Please resolve the merge conflict and those points above and we can check again the new image together. Thanks!

rhcarvalho commented:
@omron93 what are we missing to get this merged?

omron93 commented:
@rhcarvalho Hi, replication should be working now. Please test.

One change is that post-hook pod should be also listed by a service. To allow other pods to add itself to a replica set. I know that it is not standard to list hook pod in service. So I added a new service: mongodb-deploy - this service was added to list post-hook pod -> @rhcarvalho How to do it better? I wonder that now in only works for `#1` deployment :-(  ). Or it isn't a problem to add post-hook pod to mongodb service?

I hope this redesign should fix a problems with DNS/dig resolving.

rhcarvalho commented:
@omron93 please ping me here when you get this updated ;)

Commit:
Use only one kubernetes service.

omron93 commented:
@rhcarvalho It is updated.

Please look at it to be able to merge it. And to avoid a duplication of work.

Before the merge I would like to also apply these changes to 2.4 version.

omron93 commented:
@php-coder Please look at this PR. I hope it will fix a lot of things you are now working on (your PRs there)...

rhcarvalho commented:
@php-coder could you please review this tomorrow? We might take a look at this together as well.

@omron93 I would like to get the updates from https://github.com/openshift/mongodb/pull/131 in here as well, would that be possible?

omron93 commented:
@rhcarvalho Of course. So https://google.github.io/styleguide/shell.xml is a preferred code style?

mfojtik commented:
@omron93 there is an open issue about bash styling here: https://github.com/openshift/origin/pull/6598

(but it is a subset of the google shell guide, which I think is safe to follow ;-)

rhcarvalho commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,91 @@
+function cache_container_addr() {
+  echo -n "=> Waiting for container IP address ..."
+  for i in $(seq $MAX_ATTEMPTS); do
+    result=$(ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | cut -f1 -d'/')
```
@omron93 this function also changed recently, could you please update?

Here is the new code:
https://github.com/openshift/mongodb/blob/master/2.4/root/usr/share/container-scripts/mongodb/common.sh#L26-L39

rhcarvalho commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,91 @@
+}
+
+# replset_addr return the address of the current replSet
+# NOTE: this function hopes that some replset member use standart port
```
This note sounds scary... I couldn't find this function in master, but I recall there is some (possibly bad) assumption about the port for the replication example.

I would prefer an open issue for us to solve that problem than an easy-to-ignore comment, what do you say?

rhcarvalho commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,91 @@
+# $2 - host where to connect (replset_addr by default)
+function mongo_wait_replset() {
+  # If there is no PRIMARY yet, rs.isMaster().primary returns "undefined"
+  for i in $(seq 1$MAX_ATTEMPTS); do
```
Looks like a typo here, `1$MAX_ATTEMPTS`, there should be no `1`, right?

rhcarvalho replied:
Please make `i` local, in a line by itself before the for loop.

rhcarvalho commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,91 @@
+    # Test connection to replica set
+    local set=$(mongo admin ${1:-} --host ${2:-$(replset_addr)} --quiet --eval "rs.isMaster().setName;" | tail -n 1)
+    # If there is no PRIMARY yet, rs.isMaster().primary returns "undefined"
+    local primary=$(mongo admin ${1:-} --host ${2:-$(replset_addr)} --quiet --eval "rs.isMaster().primary;" | tail -n 1)
```
`local` declarations and assignments should go in separate lines, specially when there's a subshell / evaluation on the right side. Here is the explanation why: https://google.github.io/styleguide/shell.xml?showone=Use_Local_Variables#Use_Local_Variables

rhcarvalho commented:
@omron93 we're still missing the "headers" in every file setting `errexit`, `nounset`, `pipefail`...

To be sincere, as we move along with this I don't see much of the value of this rewrite. It's a big change, there's a lot of work to catch up with recent "fixes" to bend around Bash gotchas... and God only knows what new bugs we'll be introducing (like the possible typo in the for loop I spotted).

Not telling you to give it up. I believe you will still want these changes, as we have chatted before this PR carries the benefit of splitting replication from non-replication code, so we'll need to make sure we update all of the code you moved to match the recent state of master ;)

Commit:
Use shell coding style.

omron93 commented:
@rhcarvalho Yes, I was working on it. Now the code should be according shell coding style.

Yes, main goal what I want to achieve by this PR is to split replication and noreplication part. Also I thing it is a better approach to use MongoDB style to connect to replSet - rs_name/host1,host2,... and do not reinvent the wheel with finding master. User of this MongoDB image will probably use this method too. So they theoretically could use the same implementation.
`docker run -ti centos/mongodb-26-centos7 bash -c 'source ${CONTAINER_SCRIPTS_PATH}/replset-functions.sh; mongo admin -u admin -p $MONGODB_ADMIN_PASSWORD --host $(replset_addr)'`

rhcarvalho commented:
> Also I thing it is a better approach to use MongoDB style to connect to replSet - rs_name/host1,host2,...

Yep, this is correct. I had a branch with that change against master as well, but haven't proposed a PR because it seems that sometimes we do need the other style.

Actually I had the dumb `--host rs0/localhost`, but that's not good, it doesn't serve our purposes before the replica set is configured.

I wonder if replica set name + list of hosts (that are not connected as a replica set yet) will work? Have you tried it? I haven't had a chance to evaluate that yet.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+
+set -o errexit
+set -o nounset
+set -o pipefail
```
Probably it's better to set these options before `source`ing other files.

omron93 replied:
You are right. I will fix it.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+
+  echo "=> Successfully added to the replSet ..."
+
+fi
```
It's up to you, but I would rather extract both of the `if`'s branches into separate functions.

omron93 replied:
I have nothing against it... but I don't understand why? This peaces of code won't be used in another place...

rhcarvalho replied:
I don't know what @php-coder had in mind, but for me functions are more than being able to call them in multiple places, they can give a name to a procedure and make reading and understanding the code easier.

There are different schools of thought though. One might be more focused on "coding for the computer", others "coding for others to read", and others something else ;)

omron93 replied:
> I don't know what @php-coder had in mind, but for me functions are more than being able to call them in multiple places, they can give a name to a procedure and make reading and understanding the code easier.

The code is separated in a separate file (to note - it same to container-scripts/mongodb/initiate_replica.sh file - should everything be "separated" in a new function?).

IMPOV a separation of almost all code in a file to make a function from it to "hide" IF-ELSE statement to contains only one line in each branch with function call is not the way to implement "coding for others to read" :-(

However if this is a style in OpenShift I can do it.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+
+  # Wait for at least 3 members
+  members=0
+  while [ "${members}" -le 3 ] 2>/dev/null || [[ $? -ne 1 ]] ; do
```
Why do we need `2>/dev/null` here? Who prints to stderr here?  Overall, I can't fully understand this line.. :-(

omron93 replied:
If `members` contains non-numeric value test will print an error message therefore there is `2>/dev/null`. And in that case we still wants to wait so the `[[ $? -ne 1 ]]` is there to indicate that the previous test failed with $?=2 (contains non-numeric value)...

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+  mongo_create_user "-u admin -p ${MONGODB_ADMIN_PASSWORD}"
+
+  # Add one pod to this replSet to be able to address it with mongodb service
+  member_addr=$(echo "$(endpoints)" | head -n 1)
```
Probably, you could remove `echo` here..

omron93 replied:
OK, thanks. I will remove it.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+  done
+
+  # Exit this pod
+  kill $2
```
I would rather add human-readable name to this variable, like this:

``` bash
process_id=$2
...
kill ${process_id}
```

omron93 replied:
OK

rhcarvalho replied:
+1 for most if not all places using $1, $2, $3... it is hard to read the middle of a function and figure out what $3 is supposed to mean.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+set -o nounset
+set -o pipefail
+
+set -x
```
Why do we need it? Is it temporary, for debugging, or permanent?

omron93 replied:
Sorry, it is only for debuging. Removed.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,97 @@
+
+# setup_keyfile fixes the bug in mounting the Kubernetes 'Secret' volume that
+# mounts the secret files with 'too open' permissions.
+function setup_keyfile() {
```
Comment looks like outdated. Maybe it fixes bug too, but the main purpose is to create keyfile for MongoDB.

omron93 replied:
This is already in master branch. So I copied it.

I agree with you. Changed.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-pre-init.sh:
```diff
@@ -0,0 +1,35 @@
+# This file sets some basic mongod replication variables and checks mandatory replication variables
+
+# Add parameters for replication
+export mongod_common_args+=" --oplogSize 64 --replSet ${MONGODB_REPLICA_NAME} "
```
You're using `${MONGODB_REPLICA_NAME}` here but validate it's value at the end of the file. Is it ok? Is it possible that it will fail here because of using undeclared variable?

omron93 replied:
Thanks, moved.

It was at the end of file but I moved it during bash sanitization to make it consistent with pre-init.sh. My fault.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,97 @@
+
+# replset_addr return the address of the current replSet
+function replset_addr() {
+  echo "${MONGODB_REPLICA_NAME}/$(echo $(endpoints) | sed -e 's| |:27017,|g' -e 's|$|:27017|')"
```
Probably, you could replace `echo $(endpoints)` with `endpoints`

php-coder replied:
What these sed expressions actually do?

omron93 replied:
echo removes the new lines from dig output and sed adds port number to each address and joins them with comma -> `repl_name/host1,host2,...`

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,97 @@
+    fi
+    sleep ${SLEEP_TIME}
+  done
+  echo "Failed to get Docker container IP address." && exit 1
```
Why you use `&& exit 1`? Why do you need `&&` here?

omron93 replied:
I don't know. Also copied from master branch.

I make sense to me to have `exit` on separate line. Changed.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,97 @@
+  local i
+  for i in $(seq ${MAX_ATTEMPTS}); do
+    if ip -oneline -4 addr show up scope global | grep -Eo '[0-9]{,3}(\.[0-9]{,3}){3}' > "${HOME}"/.address; then
+      echo " $(mongo_addr)"
```
Why do we need leading space here?

omron93 replied:
To separate comment file header and function commend. Should I remove it?

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/replset-functions.sh:
```diff
@@ -0,0 +1,97 @@
+  echo -n "=> Waiting for container IP address ..."
+  local i
+  for i in $(seq ${MAX_ATTEMPTS}); do
+    if ip -oneline -4 addr show up scope global | grep -Eo '[0-9]{,3}(\.[0-9]{,3}){3}' > "${HOME}"/.address; then
```
Is it ok that in file could be more than one IP address?

Output of this command on Vagrant:

```
10.0.2.15
10.0.2.255
172.17.0.1
```

Output of this command on my host:

```
10.34.129.229
10.34.129.255
10.200.136.146
10.200.139.255
192.168.122.1
192.168.122.255
```

omron93 replied:
It is better to use only one IP for connection via mongo shell - I think `mongo --host IP1 IP2` could cause an error. Thanks.

This is also present in master branch.

php-coder commented on 2.6/root/usr/bin/run-mongod-replication:
```diff
@@ -0,0 +1,75 @@
+# Run replset-supervisor
+${CONTAINER_SCRIPTS_PATH}/replset-supervisor.sh "${1:-}" "$$" & supervisor_pid=$!
+
+# Run `unset_env_vars` and `mongod` in a subshell because
```
Probably, this comment should be updated?

omron93 replied:
OK

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/post-init.sh:
```diff
@@ -0,0 +1,8 @@
+  unset MONGODB_USER MONGODB_PASSWORD MONGODB_DATABASE MONGODB_ADMIN_PASSWORD
+}
+
+unset_env_vars
```
I don't see the reason of having this function if it's used only once here, right after declaration.

omron93 replied:
You are right.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/init.sh:
```diff
@@ -0,0 +1,35 @@
+  mongo_reset_passwords
+fi
+
+if [[ -f ${dbpath}/.mongodb_datadir_initialized ]]; then
```
This check could be merged with `else` branch of the previous `if`. Or I missed something?

omron93 replied:
I wanted to make it more general. And complicated :-(

IF removed.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/init-functions.sh:
```diff
@@ -0,0 +1,83 @@
+  set +e
+  mongo admin ${1:-} --host ${2:-"localhost"} --eval "db.createUser({user: 'admin', pwd: '${MONGODB_ADMIN_PASSWORD}', roles: ['dbAdminAnyDatabase', 'userAdminAnyDatabase' , 'readWriteAnyDatabase','clusterAdmin' ]});"
+  local result=$?
+  set -e
```
Why do we need this `set +e/-e`?

omron93 replied:
Because mongo call could fail to connect to mongod. So to be able to check it and to print error message.

Is it too complicated? Is mongo fail message enought?

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/init-functions.sh:
```diff
@@ -0,0 +1,83 @@
+# mongo_create_admin creates the MongoDB admin user with password: MONGODB_ADMIN_PASSWORD
+# $1 - login parameters for mongo (optional)
+# $2 - host where to connect (localhost by default)
```
IMHO this comments could be replaced by code:

``` bash
local mongo_login="${1:-}"
local mongo_host="${2:-localhost}"
```

omron93 replied:
No problem with this.

However in that shell coding style there is also documentation of parameters in function comment. So what is the OpenShift way?

rhcarvalho replied:
@omron93 you don't need to try to follow the style guide 100%, as the current code base doesn't.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/base-functions.sh:
```diff
@@ -0,0 +1,76 @@
+# Change value for config option in configuration file
+# $1 - option name
+# $2 - new value
+# $3 - path to config file
```
Please, replace this comments with readable code:

``` bash
local option_name="${1:-}"
local option_value="${2:-}"
local config_path="${3:-}"
```

omron93 replied:
Same as above.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/base-functions.sh:
```diff
@@ -0,0 +1,76 @@
+# $2 - new value
+# $3 - path to config file
+function update_option() {
+  [[ -z "${1:-}" || -z "${2:-}" || -z "${3:-}" || ! -r "${3:-}" ]] && return 1
```
Unfortunately there is no error messages that could help in debugging :-(

omron93 replied:
Added.

php-coder commented on 2.6/root/usr/share/container-scripts/mongodb/base-functions.sh:
```diff
@@ -0,0 +1,76 @@
+  rm ${HOME}/.tmp.conf
+
+  # Add new option into config file
+  echo "$1 = $2" >> $3
```
This code do non-atomic operation and it's possible to leave temp file in case of the error.

Also this could be replaced by sed one-liner:

``` bash
sed -i "s|^\([[:space:]]$option_name[[:space:]]=[[:space:]]\).*|\1$option_value|" $config_path
```

omron93 replied:
Yes it is. But this can't influence the next run of this function - `sed -r -e "/^\s*$1/d" $3 > ${HOME}/.tmp.conf` completely rewrite this file. And "in case of the error" the container will stop, so it doesn't matter.

Also it is not possible by this one-line (I wanted to use first) -  `sed -i` uses temp file in directory where the result file is. And mongodb linux user has no right to create this file in `/etc`, so therefore there is a need to create this temporary file in HOME.

omron93 replied:
Using a simpler sed command now. But temp file in HOME still needed...

Commit:
Some code changes from PR comments.

omron93 commented:
> Yep, this is correct. I had a branch with that change against master as well, but haven't proposed a PR because it seems that sometimes we do need the other style.

When, please?

> Actually I had the dumb --host rs0/localhost, but that's not good, it doesn't serve our purposes before the replica set is configured.

You are right, it is not working. So in that case `mongo_addr` is used in --host. However it is only at the beginning of post-hook pod run.

> I wonder if replica set name + list of hosts (that are not connected as a replica set yet) will work? Have you tried it? I haven't had a chance to evaluate that yet.

Yes, I've tried it. MongoDB only prints warning that hostXY is not part of the replicaset. It is enought that one specified and accessible host is part of the replica set.
This "functionality" is often used in this PR :-)

rhcarvalho commented on 2.6/root/usr/share/container-scripts/mongodb/replset-supervisor.sh:
```diff
@@ -0,0 +1,58 @@
+
+  # Wait for at least 3 members
+  members=0
+  while [ "${members}" -le 3 ] 2>/dev/null; do
```
Why would we hard-code waiting for 3 members?!

omron93 replied:
When should the OpenShift deployment finish? In case at least one member is ready for connection?

rhcarvalho commented:
This PR is already too complex to review, we're reviewing both moving things into new files + little and not so little improvements and changes everywhere. And we need to account for the fact that master has changed many times since the time you moved things around originally.

Would be much easier to review if the improvements were more incremental. For example, there could be many PRs fixing the implementation problems of existing functions (as we are discussing them here). Then eventually a PR splitting the code.

But since we talked that the main idea here was to split replication from non-replication code, we could move code first, with only minimal changes to keep things working if needed. We review it. Test it. Merge it. And then improve function by function.

Please don't throw away your improvements, keep them in a separate branch and we will review the other improvements in new, smaller PRs.

What do you think? Would it be possible to keep this PR only about obvious to verify code moves?

omron93 commented:
> And we need to account for the fact that master has changed many times since the time you moved things around originally.

So lets concentrate to get this PR done ;-)

> Would be much easier to review if the improvements were more incremental. For example, there could be many PRs fixing the implementation problems of existing functions (as we are discussing them here).

In this PR there are changes in functions present in current implementation:
1. wait_mongo -> joins 2 almost same function wait_for_mongo_up and wait_for_mongo_down (they even aren't so much consistent in master :-( ) - OK it should be fixed in separe PR
2. mongo_create_admin, mongo_create_user -> splits a function to create MongoDB admin and normal user + adds check of env variables - OK it should be fixed in separe PR
3. mongo_reset_passwords -> only adds check of env variables - OK it should be fixed in separe PR

In replication part there are changes only in: 
4. mongo_initiate -> due to different schema o replSet initiation; mongo_remove, mongo_add -> small changes due to different way of replSet addressing

This PR also changes:
5. move config file to /etc and chage it to not to use envsubst -> to be able to access config file in case user is not running default container commands (he runs for example bash) - this adds only some functions and only small change in run-mongod, so OK it should be fixed in separe PR
6. use MongoDB style `replSet/host1,host2,...` to address replSet -> one new function and changes in addressing replSet in several functions (see 3.)
7. split replication and non replication part of run-mongod, also make it more clean, make initiation and adding new pods more similar, not to make post-deploy-hook completely different,...

1.,2.,3. and 5. are really small and it is not so hard to propose them in separate PR if you wish. However I am not so convinced this will not break a style consistency in this PR...

Point 6. and part of 4. are quite bigger and I really think it is waste of time to try to fit it to the weird schema of run-mongod if 7. is ready!

To be honest, of course in this PR are bugs. But bugs are also in the current implementation and I think that the number of them are not bigger in this PR. Also I am convinced that this PR improves maintainability of this images and it is the first step to extend this image - to for example add support for sharding.

OK, lets check every new and changed function. We can do it systematically and it could be done in one PR.

php-coder commented:
@omron93 Could you sync your PR with the latest changes from master? I see that at least my fix for getting the IP of the primary node is missing.

Commit:
Use regexp to get PRIMARY.

omron93 commented:
@php-coder Checking if db.isMaster().primary is OK is now using your regexp.

bparees commented:
@omron93 @php-coder @rhcarvalho what's the state of this PR (aside from needing a rebase)?

rhcarvalho commented:
@bparees I think this got pretty big to test/review and was split into several follow up PRs -- several still open. Perhaps just a rebase to uncover what's here that is not in any other PR.

omron93 commented:
All separated PRs are now merged. So closing this PR.
